psydac/linalg/solvers.py

```diff
@@ -1725,3 +1728,260 @@ def solve(self, b, out=None):
+                assert value*0 == 0, "tol must be a real number"
+                assert value > 0, "tol must be positive"
+            elif key == 'maxiter':
+                assert value is not None, "maxiter may not be None"
+                assert isinstance(value, int), "maxiter must be an int"
+                assert value > 0, "maxiter must be positive"
+            elif key == 'verbose':
+                assert value is not None, "verbose may not be None"
+                assert isinstance(value, bool), "verbose must be a bool"
+            else:
+                raise ValueError(f"Key '{key}' not understood. See self._options for allowed keys.")
+
+    def transpose(self, conjugate=False):
+        At = self._A.transpose(conjugate=conjugate)
+        solver = self._solver
+        options = self._options
+        return inverse(At, solver, **options)
+
+    def solve(self, b, out=None):
+        """
+        Generalized minimum residual algorithm for solving linear system Ax=b.
+        Implementation from Wikipedia
+        Info can be accessed using get_info(), see :func:~`basic.InverseLinearOperator.get_info`.
+
+        Parameters
+        ----------
+        b : psydac.linalg.basic.Vector
+            Right-hand-side vector of linear system Ax = b. Individual entries b[i] need
+            not be accessed, but b has 'shape' attribute and provides 'copy()' and
+            'dot(p)' functions (dot(p) is the vector inner product b*p ); moreover,
+            scalar multiplication and sum operations are available.
+
+        out : psydac.linalg.basic.Vector | NoneType
+            The output vector, or None (optional).
+
+        Results
+        -------
+        x : psydac.linalg.basic.Vector
+            Converged solution.
+
+        info : dict
+            Dictionary containing convergence information:
+            - 'niter'    = (int) number of iterations
+            - 'success'  = (boolean) whether convergence criteria have been met
+            - 'res_norm' = (float) 2-norm of residual vector r = A*x - b.
+
+        References
+        ----------
+
+        """
+
+        A = self._A
+        domain = self._domain
+        codomain = self._codomain
+        options = self._options
+        x0 = options["x0"]
+        tol = options["tol"]
+        maxiter = options["maxiter"]
+        verbose = options["verbose"]
+        
+
+        assert isinstance(b, Vector)
+        assert b.space is domain
+
+        # First guess of solution
+        if out is not None:
+            assert isinstance(out, Vector)
+            assert out.space is codomain
+
+        x = x0.copy(out=out)
+        
+        # Extract local storage
+        r = self._tmps["r"]
+        v = self._tmps["v"]
+
+        # Internal objects of GMRES
+        H = self._H
+        Q = []
+        beta = []
+        sn = []
+        cn = []
+
+        # First values
+
+        A.dot( x , out=r)
+
+        r += b - r 
```

Comment:
Did the previous dot product `A.dot( x , out=r)` is not useless since at the end we only have `b` store in `r` ?

Reply:
Also we should optimize the `r += b - r` line by only doing internal operation

Reply:
This was an error. The first residual vector must be `r = b - A.dot(x)`. 
An extra temporary is necessary to save the content of `A.dot(x)` and it cannot be saved in `b`, because that would rewrite it.